src/node_crypto.cc

```diff
@@ -2506,8 +2506,31 @@ void SSLWrap<Base>::GetALPNNegotiatedProto(
   if (!alpn_proto)
     return args.GetReturnValue().Set(false);
 
-  args.GetReturnValue().Set(
-      OneByteString(args.GetIsolate(), alpn_proto, alpn_proto_len));
+  Local<String> result;
+  switch (alpn_proto_len) {
+    case sizeof("h2") - 1:
+      if (alpn_proto[0] == 'h' && alpn_proto[1] == '2')
+        result = w->env()->h2_string();
+      break;
+
+    case sizeof("http/1.1") - 1:
+      if (alpn_proto[0] == 'h' &&
+          alpn_proto[1] == 't' &&
+          alpn_proto[2] == 't' &&
+          alpn_proto[3] == 'p' &&
+          alpn_proto[4] == '/' &&
+          alpn_proto[5] == '1' &&
+          alpn_proto[6] == '.' &&
+          alpn_proto[7] == '1') {
```

Comment:
Btw, it looks like modern versions of gcc and clang can reduce a `memcmp()` against `http/1.1` to a single comparison because it happens to be 8 bytes in length (or 2 comparisons for 32-bit platforms), whereas a chained if seems to yield 7 individual branches: https://godbolt.org/z/U9KRY2

Reply:
I'll make that change if you think it's a good idea. I actually had them as memcmps first. :)

The reason I went with open-coding is because it's not always clear to me when gcc and clang decide to apply the optimization. I looked it up in clang and there are a bunch of conditions involved.

Reply:
Yeah, I’d go with memcmp – it’s also nice in that it is a bit less verbose. (And it probably doesn’t matter too much anyway?)